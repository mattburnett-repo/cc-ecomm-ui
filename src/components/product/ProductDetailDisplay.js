
import { useHistory } from 'react-router-dom'
import { useDispatch } from 'react-redux'

import ItemQuantity from '../../widgets/ItemQuantity'
import { addItemToCurrentCart } from '../../features/carts/currentCartActions'

import { StyledProductDetailDisplay } from '../styles/ProductDetailDisplay.styled'

// USE THIS COMPONENT TO BREAK OUT AND SHOW INFO ABOUT ONE PRODUCT
//      use ProductDisplay to display products in ProductsListingDisplay

export default function ProductDetailDisplay ( props ) {
    const { item } = props

    const history = useHistory()
    const dispatch = useDispatch()

    return (
        <>
            <header>
            <h4>Product Detail Display</h4>
            </header>
        
            <StyledProductDetailDisplay>
                <div id="productImage" role="presentation" aria-label="product-image">
                    <div role="presentation" aria-label='product' key={item.id}></div>
                    <div role="img" id="product-image-url" aria-label='product-image-url'> 
                        <img src={item.image_url} alt={item.name} id={item.id} />
                    </div>                
                </div>
                <div id="productInfo" role="presentation" aria-label='product-info'>
                    <label htmlFor='product-name'>Product Name:</label>
                    <div id="product-name" role='presentation' aria-label='product-name'>{item.name}</div>
                    <label htmlFor='product-description'>Product Description:</label>
                    <div id="product-description" role='presentation' aria-label='product-description'>{item.description}</div>

                    <div role="presentation" aria-label='product-order-controls'>
                        <label htmlFor='product-price'>Price:</label>
                        <div role='presentation' aria-label='product-price'>{item.price}</div>

                        <div role="presentation" aria-label='product-item-quantity'>
                            <ItemQuantity itemId={item.id}/>
                        </div>
                        <button aria-label="add-to-cart" onClick={() => dispatch(addItemToCurrentCart(item))}>Add To Cart</button>
                        <button aria-label="go-back-button" onClick={() => history.goBack()}>Go Back</button>
                    </div>              
                </div>
            </StyledProductDetailDisplay>
        </>
    )
}