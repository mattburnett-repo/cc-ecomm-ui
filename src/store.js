import { configureStore } from '@reduxjs/toolkit';

import authReducer from './features/auth/authSlice';
import productCategoryReducer from './features/productCategory/productCategorySlice'
import productsListingReducer from './features/products/productsSlice'
import ordersReducer from './features/orders/ordersSlice'
import cartsReducer from './features/carts/cartsSlice'

function rootReducer() {
    return {
        reducer: {
            auth: authReducer,
            productCategories: productCategoryReducer,
            orders: ordersReducer,
            carts: cartsReducer,
            products: productsListingReducer,
        }    
    }
}

const store = configureStore(rootReducer());

export default store;