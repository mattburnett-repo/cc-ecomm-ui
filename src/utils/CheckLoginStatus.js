
import { useSelector } from 'react-redux'
import { useHistory } from 'react-router-dom'

import { selectIsAuthenticated } from '../features/auth/authSlice'

export default function CheckLoginStatus () {
    // eslint-disable-next-line
    const isLoggedIn = useSelector(selectIsAuthenticated);

    const history = useHistory()

    if(!isLoggedIn) {
        history.push('/login')

        return false
    } else {
        return true;
    }
}        
