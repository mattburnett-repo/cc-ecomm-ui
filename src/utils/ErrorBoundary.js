import React from 'react'

export default class ErrorBoundary extends React.Component {
    constructor(props) {
      super(props);
      this.state = { hasError: false };
    }
  
    static getDerivedStateFromError(error) {
      // Update state to trigger fallback UI
      return { hasError: true };
    }
  
    componentDidCatch(error, errorInfo) {
        // Log error or send logging data to log management tool
        //   logErrorToMyService(error, errorInfo);
        console.log('ErrorBoundary ', errorInfo)
    }
  
    render() {
      if (this.state.hasError) {
        // Render fallback UI
        return <h1>There was an error. Sorry about that; we're working on it.</h1>;
      }
  
      return this.props.children; 
    }
  }

