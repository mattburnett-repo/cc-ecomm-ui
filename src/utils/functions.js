
// USAGE: formatAsCurrency.format(3.005); // "3.01"
const formatAsCurrency = new Intl.NumberFormat('en-US', {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
});

const formatDate =   new Intl.DateTimeFormat("en-US", {
    year: "numeric",
    month: "long",
    day: "2-digit"
});

module.exports = {
    formatAsCurrency: formatAsCurrency,
    formatDate: formatDate
}