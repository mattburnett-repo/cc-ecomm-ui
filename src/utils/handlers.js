
// handler function declarations

function handleSaveCart(e) { 
    alert('handleSaveCart inside utils ')
} 

function handleCheckout(e) { 
    alert('handleCheckout inside utils ')
} 

function handleQuantityChange(e, itemId) { 
    alert('handleQuantityChange inside utils ' + e.target.value + ' id: ' + itemId)
} 

function handleSearch(e) {
    e.preventDefault()
    // let searchTerms = e.target.value
    alert('handleSearch for search term/s ' + e.target.searchTerms.value + '. \nneed to figure out how to get dispatch() to work from here')
}

function handleOrderClick(e) {
    e.preventDefault()

    let order_id = e.target.order_id.value
    alert('go to order detail component ' + order_id)
}

function handleHomeClick() {
    console.log('handleHomeClick')
}
function handleShowCartClick() {
    console.log('handleShowCartClick')
}
function handleCheckoutClick() {
    console.log('handleCheckoutClick')
}
function handleCancelClick() {
    console.log('handleCanceClick')
}

module.exports = {
    handleSaveCart: handleSaveCart,
    handleHomeClick: handleHomeClick,
    handleShowCartClick: handleShowCartClick,
    handleCheckoutClick: handleCheckoutClick,
    handleCancelClick: handleCancelClick,
    handleQuantityChange: handleQuantityChange,
    handleCheckout: handleCheckout,
    handleSearch: handleSearch,
    handleOrderClick: handleOrderClick
} // end module.exports