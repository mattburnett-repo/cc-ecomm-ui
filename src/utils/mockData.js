// https://css-tricks.com/writing-tests-for-react-applications-using-jest-and-enzyme/

module.exports = {
    // someday I will figure out how to do this correctly...
    // getUserData: jest.fn(() => {
    //   return Promise.resolve({
    //   data: [
    //     {
    //       id: 1,
    //       username: 'asdf',
    //       email: 'asdf@asdf.com',
    //       password: 'asdf'
    //     }
    //   ]
    //   })
    // }),
    mockCartsData:[
        {
          "cart": {
              "cart_id": 1,
              "user_id": 1,
              "cart_items": [
                  {
                      "product_id": 1,
                      "product_name": "hedgehog",
                      "product_price": 1,
                      "product_quantity": 1,
                      "product_total_price": 1
                  },
                  {
                      "product_id": 2,
                      "product_name": "maine coon cat",
                      "product_price": 2,
                      "product_quantity": 2,
                      "product_total_price": 4
                  },
                  {
                      "product_id": 3,
                      "product_name": "bulldog",
                      "product_price": 3,
                      "product_quantity": 3,
                      "product_total_price": 9
                  },
                  {
                      "product_id": 4,
                      "product_name": "bunny",
                      "product_price": 4,
                      "product_quantity": 4,
                      "product_total_price": 16
                  }
              ],
              "cart_total_price": {
                  "sum": 30,
                  "cart_id": 1
              }
          }
      }
    ],
    mockCartData: {
      cart: {
        cart_id: 1,
        user_id: 1,
        cart_items: [
          {
            product_id: 1,
            product_name: 'hedgehog',
            product_price: 1,
            product_quantity: 1,
            product_total_price: 1
          },
          {
            product_id: 2,
            product_name: 'maine coon cat',
            product_price: 2,
            product_quantity: 2,
            product_total_price: 4
          },
          {
            product_id: 3,
            product_name: 'bulldog',
            product_price: 3,
            product_quantity: 3,
            product_total_price: 9
          },
          {
            product_id: 4,
            product_name: 'bunny',
            product_price: 4,
            product_quantity: 4,
            product_total_price: 16
          }
      ],
      cart_total_price: { sum: 30, cart_id: 1 }
    }},
    mockOrderData: [{
      "order": {
          "cart": {
              "cart_id": 1,
              "cart_items": [
                  {
                      "product_id": 1,
                      "product_name": "product_name_01",
                      "product_price": 1,
                      "product_quantity": 1,
                      "product_total_price": 1
                  },
                  {
                      "product_id": 2,
                      "product_name": "product_name_02",
                      "product_price": 2,
                      "product_quantity": 2,
                      "product_total_price": 4
                  },
                  {
                      "product_id": 3,
                      "product_name": "product_name_03",
                      "product_price": 3,
                      "product_quantity": 3,
                      "product_total_price": 9
                  },
                  {
                      "product_id": 4,
                      "product_name": "product_name_04",
                      "product_price": 4,
                      "product_quantity": 4,
                      "product_total_price": 16
                  }
              ]
          },
          "user_id": 1,
          "order_id": 1,
          "order_date": null,
          "total_price": {
              "sum": 30,
              "cart_id": 1
          }
        }
      }],
    mockOrderConfirmationData:
      {
        "order_id": 1,
        "total_amount": 1234,
        "order_date": '12-Dez-2021',
        "payment_type": 'Deutschmarks',
        "delivery_address": '123 Katzendamm, 90210 Berlin DE'
      },
    mockProductCategories: [
        // {
        //   "category_id": 0,
        //   "description": "Please select a product category"
        // },
        {
          "category_id": 4,
          "description": "Auto"
        },
        {
            "category_id": 3,
            "description": "Health"
        },
        {
            "category_id": 1,
            "description": "Pets"
        },
        {
            "category_id": 2,
            "description": "Tech"
        }
      ],
    mockProduct:[
        {
          "id": 2,
          "category_id": 2,
          "name": "product_name_02",
          "description": "product_desc_02",
          "image_url": "https://cdn.pixabay.com/photo/2015/03/27/13/16/maine-coon-694730_1280.jpg",
          "price": "2.00"
        },      
      ],
    mockProducts: [
        {
            "id": 1,
            "category_id": 1,
            "name": "product_name_01",
            "description": "product_desc_01",
            "image_url": "https://pixabay.com/images/id-71282/",
            "price": "1.00"
        },
        {
            "id": 2,
            "category_id": 2,
            "name": "product_name_02",
            "description": "product_desc_02",
            "image_url": "http://www.image.com/item/2",
            "price": "2.00"
        },
        {
            "id": 3,
            "category_id": 3,
            "name": "product_name_03",
            "description": "product_desc_03",
            "image_url": "http://www.image.com/item/3",
            "price": "3.00"
        },
        {
            "id": 4,
            "category_id": 4,
            "name": "product_name_04",
            "description": "product_desc_04",
            "image_url": "http://www.image.com/item/4",
            "price": "4.00"
        }
    ],
    mockUserData:[
      {
        "id": 4,
        "user_name": "asdf",
        "password": "$2b$10$OsTX1L59rYgsmAnsrJsXoO3RH3RSfLgEF53YBB3723B9eoELtUa.a",
        "email": "asdf@asdf.com",
        "google_id": null,
        "google_display_name": null,
        "google_first_name": null,
        "google_last_name": null,
        "google_image": null
      }
    ]
  } // end module.exports