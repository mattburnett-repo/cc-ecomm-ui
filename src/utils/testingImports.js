// export default function TestingImports() {
//     return (
//         import { render, screen, fireEvent } from '@testing-library/react';
//         import renderer from 'react-test-renderer';
//         import userEvent from '@testing-library/user-event';

//         import { Provider } from 'react-redux';
//         import store from '../../store'
//     )
// }