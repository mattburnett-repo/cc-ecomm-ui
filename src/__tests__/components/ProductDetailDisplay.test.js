// 20211217: Wireframe https://wireframe.cc/7nTPt0

import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { Provider } from 'react-redux';
import store from '../../store'

import ProductDetailDisplay from '../../components/product/ProductDetailDisplay';

import { mockProduct } from '../../utils/mockData'

const mockHandleAddToCart = jest.fn()

describe('<ProductDetailDisplay /> component tests', () => {
    beforeEach(() => render(<ProductDetailDisplay data={mockProduct} handleAddToCart={ mockHandleAddToCart }/>));

    it('should render a Go Back button', () => {
        let theVals = screen.getAllByRole('button', { name: /product-detail-go-back-button/i})
        expect(theVals).toHaveLength(2)
    })
    it('should display a name', () => {
        screen.getByRole('presentation', {name: /name/i});
        expect(screen.getByRole('presentation', {name: /name/i})).toHaveTextContent('glowing cube');
    });
    it('should display a product image', () => {
        screen.getByRole('img', {name: /product-image/i});
        expect(screen.getByRole('img', {name: /product-image/i})).toHaveAttribute('src', 'https://cdn.pixabay.com/photo/2017/07/10/23/45/cubes-2492010__340.jpg');
    });
    it('should display a description', () => {
        screen.getByRole('presentation', {name: /description/i});
        expect(screen.getByRole('presentation', {name: /description/i})).toHaveTextContent('prod desc 02');
    });
    it('should display a price', () => {
        screen.getByRole('presentation', {name: /price/i});
        expect(screen.getByRole('presentation', {name: /price/i})).toHaveTextContent('1.00');
    });
    it('should display a quantity input', () => {
        screen.getByRole('textbox', {name: /quantity/i});
        expect(screen.getByRole('textbox', {name: /quantity/i})).toHaveTextContent('');
    });
    // TODO: add mockHandleAddToCart test/s
    it('should display an add-to-cart button', () => {
        let theVal = screen.getByRole('button', {name: /add-to-cart/i});
        fireEvent.click(theVal)

        expect(theVal).toHaveTextContent('Add To Cart')
    });

    it('renders a snapshot', () => {
        const tree = renderer.create(<Provider store={store}><ProductDetailDisplay data={mockProduct} /></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });
}) // end component
