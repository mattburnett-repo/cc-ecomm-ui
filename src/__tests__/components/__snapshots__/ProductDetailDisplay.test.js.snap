// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetailDisplay /> component tests renders a snapshot 1`] = `
<div>
  <h3>
    ProductDetailDisplay: 
    glowing cube
  </h3>
  <div
    aria-label="product-detail-go-back"
  >
    <button
      aria-label="product-detail-go-back-button"
      onClick={[Function]}
    >
      Go Back
    </button>
  </div>
  <img
    alt="glowing cube"
    aria-label="product-image"
    src="https://cdn.pixabay.com/photo/2017/07/10/23/45/cubes-2492010__340.jpg"
  />
  <label
    htmlFor="name"
  >
    Name:
  </label>
  <div
    aria-label="name"
    id="name"
    role="presentation"
  >
    glowing cube
  </div>
  <label
    htmlFor="product-id"
  >
    ID:
  </label>
  <div
    aria-label="product-id"
    id="product-id"
    role="presentation"
  >
    10
  </div>
  <label
    htmlFor="description"
  >
    Description:
  </label>
  <div
    aria-label="description"
    id="description"
    role="presentation"
  >
    prod desc 02
  </div>
  <label
    htmlFor="price"
  >
    Price:
  </label>
  <div
    aria-label="price"
    id="price"
    role="presentation"
  >
    1.00
  </div>
  <div
    aria-label="product-item-quantity"
    role="presentation"
  >
    <label
      htmlFor="item-quantity-selector"
    >
      Quantity:
    </label>
    <select
      aria-label="item-quantity-selector"
      defaultValue={0}
      id="item-quantity-selector"
      name="itemQuantity"
      onChange={[Function]}
      role="presentation"
    >
      <option
        value={1}
      >
        1
      </option>
      <option
        value={2}
      >
        2
      </option>
      <option
        value={3}
      >
        3
      </option>
      <option
        value={4}
      >
        4
      </option>
      <option
        value={5}
      >
        5
      </option>
      <option
        value={6}
      >
        6
      </option>
      <option
        value={7}
      >
        7
      </option>
      <option
        value={8}
      >
        8
      </option>
      <option
        value={9}
      >
        9
      </option>
      <option
        value={10}
      >
        10
      </option>
    </select>
  </div>
  <button
    aria-label="add-to-cart"
    onClick={[Function]}
  >
    Add To Cart
  </button>
  <div
    aria-label="product-detail-go-back"
  >
    <button
      aria-label="product-detail-go-back-button"
      onClick={[Function]}
    >
      Go Back
    </button>
  </div>
</div>
`;
